File: java_solutions/85.java
import java.util.* ; 
class Solution {
    public int maximalRectangle(char[][] matrix) {
        var bar_height = new int [matrix.length][matrix[0].length] ;
        for(var col = 0 ; col < matrix[0].length ; col++){
            for(var row = matrix.length-1 ; row >= 0 ;row--){
                if(row+1 < matrix.length){
                    bar_height[row][col] += bar_height[row+1][col] ; 
                }
                if(matrix[row][col] == '1'){
                    bar_height[row][col]++;
                }
                else{
                    bar_height[row][col] = 0 ;
                }
            }
        }
        // for(var row = 0 ;  row < matrix.length ; row++){
        //     for(var col = 0 ; col < matrix[0].length ; col++){
        //         System.out.print(bar_height[row][col]+" ");
        //     }
        //     System.out.println();
        // }
        Deque<Integer> bar_stack = new ArrayDeque<Integer>() ; 
        var L = new int[matrix[0].length] ; 
        var R = new int [matrix[0].length] ;
        int rec_area_max = 0 ;
        for(var row = 0 ; row  < matrix.length ; row++){
            bar_stack.clear();
            for(var col = 0 ;col < matrix[0].length ; col++){
                while(!bar_stack.isEmpty() && bar_height[row][bar_stack.peekLast()] >= bar_height[row][col]){
                    bar_stack.pollLast();
                }
                if(bar_stack.isEmpty()){
                    L[col] = -1 ;
                }
                else{
                    L[col] = bar_stack.peekLast() ;
                }
                bar_stack.offerLast(col);
            }
            bar_stack.clear() ;
            for(var col = matrix[0].length-1; col >= 0 ;col--){
                while(!bar_stack.isEmpty() && bar_height[row][bar_stack.peekLast()] >= bar_height[row][col]){
                    bar_stack.pollLast();
                }
                if(bar_stack.isEmpty()){
                    R[col] = matrix[0].length ;
                }
                else{
                    R[col] = bar_stack.peekLast() ;
                }
                bar_stack.offerLast(col);
            }
            for(var col = 0 ; col < matrix[0].length ; col++){
                // System.out.print(col+"(L: " + L[col] + " R: "+ R[col]+") ");
                rec_area_max = Math.max(rec_area_max , (R[col] - L[col] - 1) * bar_height[row][col]);
            }
            // System.out.println(rec_area_max);
            
        }
        
        return rec_area_max ; 
        
    }
}

File: rust_solutions/85.rs
impl Solution {
    pub fn maximal_rectangle(matrix: Vec<Vec<char>>) -> i32 {
        use std::cmp::max ;
        let mut bar_height = vec![vec![0;matrix[0].len()];matrix.len()] ;
        for col in 0..matrix[0].len() {
            for row in (0..matrix.len()).rev() {
                if row + 1 < matrix.len() {
                    bar_height[row][col] += bar_height[row+1][col] ;
                }
                match matrix[row][col] {
                    '1' => {bar_height[row][col] += 1;} ,
                    '0' => {bar_height[row][col] = 0;} ,
                    _ => () 
                };
            }
        }
        
        let mut bar_stack = Vec::<usize>::new() ;
        let mut L = vec![0 as usize ; matrix[0].len()];
        let mut rec_area = 0 ;
        for row in 0..matrix.len() {
            bar_stack.clear() ; 
            for col in 1..(matrix[0].len()+1) {
                while let Some(&idx) = bar_stack.last() {
                    if bar_height[row][idx-1] >= bar_height[row][col-1] {
                        bar_stack.pop();
                    }
                    else{
                        break;
                    }
                }
                L[col-1] = match bar_stack.len() {
                    0 => 0 ,
                    _ => *bar_stack.last().unwrap()
                };
                bar_stack.push(col) ;
            }
            bar_stack.clear() ;
            
            for col in (1..(matrix[0].len()+1)).rev() {
                while let Some(&idx) = bar_stack.last() {
                    if bar_height[row][idx-1] >= bar_height[row][col-1] {
                        bar_stack.pop() ;
                    }
                    else{
                        break ;
                    }
                }
                let Ri = match bar_stack.len() {
                    0 => matrix[0].len()+1 , 
                    _ => *bar_stack.last().unwrap()
                };
                bar_stack.push(col);
                rec_area = max(rec_area , ((Ri - L[col-1] -1 ) * bar_height[row][col-1]) as i32);
            }
        }
        
        rec_area 
    }
}
